### PR TITLE
Fix expiring masked CSRF cookie

### PR DESCRIPTION
**Pull request by reggieriser**
## Description

Our current masked CSRF cookie expires after 15 minutes in our deployed environments and does not appear to get regenerated or extended within the same browser session.  When the cookie expires, the system starts to throw CSRF errors.  This PR addresses several aspects of our CSRF cookie handling:

- To avoid expiring CSRF cookies, makes `masked_gorilla_csrf` cookie a (browser) session cookie, just like the base `_gorilla_csrf` cookie
- Generates a new `masked_gorilla_csrf` cookie per request (like what we were previously doing) to mitigate against the BREACH attack (http://breachattack.com/)
- Deletes both CSRF cookies upon logout (we were previously wiping the `Set-Cookie` header, but that doesn't force the browser to delete the existing cookies)
- ~~Adds an error handler to the CSRF middleware to log any CSRF issues~~

## Reviewer Notes

I have marked this as a `WIP` because I still want to deploy this branch to experimental for testing, but wanted to go ahead and get some eyes on it.

## Setup

Run the app normally and try logging in and logging out, both with dev local login and login.gov login.  Follow the cookies in your browser's dev tools panel.

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/164812764) for this change
* [Design notes](https://github.com/gorilla/csrf#design-notes) from the Gorilla CSRF library explaining the approach

### CODE REVIEW

File: pkg/auth/cookie.go [hunk 1]
```diff
@@ -136,14 +138,30 @@ func WriteMaskedCSRFCookie(w http.ResponseWriter, csrfToken string, noSessionTim
...
+	DeleteCookie(w, GorillaCSRFToken)
+}
+
+// CSRFErrorHandler handles errors that occur within the CSRF library
+func CSRFErrorHandler(logger Logger, useSecureCookie bool) http.Handler {
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		msg := fmt.Sprintf("%s - %s",
+			http.StatusText(http.StatusForbidden), csrf.FailureReason(r))
+		http.Error(w, msg, http.StatusForbidden)
+		logger.Error(msg)
```
Comment (chrisgilmerproj) [inline pkg/auth/cookie.go]: Are you just logging here? Looks nearly the same as this code: https://github.com/gorilla/csrf/blob/master/csrf.go#L272-L279

What does this buy us other than logging?

File: pkg/auth/cookie.go [hunk 2]
```diff
@@ -136,14 +138,30 @@ func WriteMaskedCSRFCookie(w http.ResponseWriter, csrfToken string, noSessionTim
...
 // MaskedCSRFMiddleware handles setting the CSRF Token cookie
-func MaskedCSRFMiddleware(logger Logger, noSessionTimeout bool, useSecureCookie bool) func(next http.Handler) http.Handler {
+func MaskedCSRFMiddleware(logger Logger, useSecureCookie bool) func(next http.Handler) http.Handler {
 	return func(next http.Handler) http.Handler {
 		mw := func(w http.ResponseWriter, r *http.Request) {
-			// Write a CSRF cookie if none exists
-			if _, err := GetCookie(MaskedGorillaCSRFToken, r); err != nil {
-				WriteMaskedCSRFCookie(w, csrf.Token(r), noSessionTimeout, logger, useSecureCookie)
-			}
+			// Write a masked CSRF cookie (creates a new one with each request)
```
Comment (chrisgilmerproj) [inline pkg/auth/cookie.go]: Can you add a more detailed note about why we decided to do it with each request? So another me doesn't come along and try to write it only once?

File: pkg/auth/cookie.go (hunk 1 above)
Comment (reggieriser) [reply pkg/auth/cookie.go]: Nothing really.  At one time @donaldthai was contemplating regenerating the CSRF token here, but we're not doing that since we regenerate with every request now.  Do we need an explicit log?  If not, we can get rid of this unless @donaldthai had other plans.

File: pkg/auth/cookie.go (hunk 2 above)
Comment (reggieriser) [reply pkg/auth/cookie.go]: Good point.  Added a quote from the gorilla/csrf docs and a link.

File: pkg/auth/cookie.go (hunk 1 above)
Comment (chrisgilmerproj) [reply pkg/auth/cookie.go]: If we want to use it to catch and regen an expired CSRF cookie then I'd say go for it. Otherwise we really don't need the logging.

File: pkg/auth/authentication/auth.go [hunk 3]
```diff
@@ -121,6 +121,7 @@ func (h LogoutHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 			session.IDToken = ""
 			session.UserID = uuid.Nil
 			auth.WriteSessionCookie(w, session, h.clientAuthSecretKey, h.noSessionTimeout, h.logger, h.useSecureCookie)
```
Comment (chrisgilmerproj) [inline pkg/auth/authentication/auth.go]: Should we just make this cleaner and delete the session cookie? Then we wouldn't have to mess with the session to figure out our deletion policy.  I'm not going to block on this and I'm fine if we decide to make a change in a separate PR.

**Comment by reggieriser** [comment]
Deployed to experimental and tested in Firefox, Safari, Chrome, and IE.  Didn't run into any CSRF issues except when I intentionally caused one by deleting the `masked_gorilla_csrf` cookie.

File: pkg/auth/cookie.go (hunk 1 above)
Comment (reggieriser) [reply pkg/auth/cookie.go]: I've removed the custom error handler, so we'll use the default that you linked to above.

File: pkg/auth/authentication/auth.go (hunk 3 above)
Comment (reggieriser) [reply pkg/auth/authentication/auth.go]: Added a separate story for this: https://www.pivotaltracker.com/story/show/164909027
